test/e2e/deployment.go

```diff
@@ -374,28 +353,62 @@ func (d *Deployment) WaitForEnvoyUpdated() error {
 }
 
 func (d *Deployment) waitForEnvoyDaemonSetUpdated() error {
-	daemonSetUpdated := func() (bool, error) {
-		tempDS := new(apps_v1.DaemonSet)
-		if err := d.client.Get(context.TODO(), client.ObjectKeyFromObject(d.EnvoyDaemonSet), tempDS); err != nil {
+	labelSelectAppEnvoy := labels.SelectorFromSet(d.EnvoyDaemonSet.Spec.Selector.MatchLabels)
+	updatedPods := func() (bool, error) {
+		ds := &apps_v1.DaemonSet{}
+		if err := d.client.Get(context.TODO(), types.NamespacedName{Name: d.EnvoyDaemonSet.Name, Namespace: d.EnvoyDaemonSet.Namespace}, ds); err != nil {
 			return false, err
 		}
-		return tempDS.Status.ObservedGeneration == tempDS.Generation &&
-			tempDS.Status.NumberAvailable > 0 &&
-			tempDS.Status.NumberAvailable == tempDS.Status.DesiredNumberScheduled &&
-			tempDS.Status.UpdatedNumberScheduled == tempDS.Status.DesiredNumberScheduled, nil
+		updatedPods := int(ds.Status.DesiredNumberScheduled)
+		if len(ds.Spec.Template.Spec.Containers) > 1 {
+			updatedPods = d.getUpdatedPods(labelSelectAppEnvoy, d.EnvoyDaemonSet.Namespace)
+		}
+		return updatedPods == int(ds.Status.DesiredNumberScheduled) &&
+			ds.Status.NumberReady > 0, nil
 	}
-	return wait.PollImmediate(time.Millisecond*50, time.Minute*3, daemonSetUpdated)
+	return wait.PollImmediate(time.Millisecond*50, time.Minute*3, updatedPods)
 }
 
 func (d *Deployment) waitForEnvoyDeploymentUpdated() error {
-	deploymentUpdated := func() (bool, error) {
-		tempDeploy := new(apps_v1.Deployment)
-		if err := d.client.Get(context.TODO(), client.ObjectKeyFromObject(d.EnvoyDeployment), tempDeploy); err != nil {
-			return false, err
+	labelSelectAppEnvoy := labels.SelectorFromSet(d.EnvoyDeployment.Spec.Selector.MatchLabels)
+	updatedPods := func() (bool, error) {
+		updatedPods := d.getUpdatedPods(labelSelectAppEnvoy, d.EnvoyDeployment.Namespace)
+		return updatedPods == int(*d.EnvoyDeployment.Spec.Replicas) &&
+			updatedPods == int(d.EnvoyDeployment.Status.ReadyReplicas) &&
+			int(d.EnvoyDeployment.Status.UnavailableReplicas) == 0, nil
+	}
+	return wait.PollImmediate(time.Millisecond*50, time.Minute*3, updatedPods)
+}
+
+func (d *Deployment) getUpdatedPods(labelSelector labels.Selector, namespace string) int {
+	contourPodImage := d.ContourDeployment.Spec.Template.Spec.Containers[0].Image
```

Comment:
Looks like we'll need to pass the desired image in as an arg, since it'll be the Envoy image when used within `WaitForEnvoy...` method

Reply:
Not quite, it's supposed to check the contour image (which exists as a sidecar in envoy as well) because that's the main image that changes after the upgrade.

Reply:
Ahh, I see. That works!